x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/queries/SparseVectorQueryBuilder.java

```diff
@@ -0,0 +1,369 @@
+import org.elasticsearch.xpack.core.ml.inference.results.TextExpansionResults.VectorDimension;
+import org.elasticsearch.xpack.core.ml.inference.results.WarningInferenceResults;
+import org.elasticsearch.xpack.core.ml.inference.trainedmodel.TextExpansionConfigUpdate;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+import static org.elasticsearch.xpack.core.ClientHelper.ML_ORIGIN;
+import static org.elasticsearch.xpack.core.ClientHelper.executeAsyncWithOrigin;
+
+public class SparseVectorQueryBuilder extends AbstractQueryBuilder<SparseVectorQueryBuilder> {
+
+    public static final String NAME = "sparse_vector";
+    public static final ParseField PRUNING_CONFIG = new ParseField("pruning_config");
+    public static final ParseField MODEL_TEXT = new ParseField("model_text");
+    public static final ParseField MODEL_ID = new ParseField("model_id");
+    public static final ParseField VECTOR_DIMENSIONS = new ParseField("vector_dimensions");
+
+    private final String fieldName;
+    private final String modelText;
+    private final String modelId;
+    private final List<VectorDimension> vectorDimensions;
+    private SetOnce<TextExpansionResults> vectorDimensionsSupplier;
+    private final TokenPruningConfig tokenPruningConfig;
+
+    public SparseVectorQueryBuilder(
+        String fieldName,
+        String modelText,
+        @Nullable String modelId,
+        @Nullable List<VectorDimension> vectorDimensions
+    ) {
+        this(fieldName, modelText, modelId, vectorDimensions, null);
+    }
+
+    public SparseVectorQueryBuilder(
+        String fieldName,
+        String modelText,
+        @Nullable String modelId,
+        @Nullable List<VectorDimension> vectorDimensions,
+        @Nullable TokenPruningConfig tokenPruningConfig
+    ) {
+        if (fieldName == null) {
+            throw new IllegalArgumentException("[" + NAME + "] requires a fieldName");
+        }
+
+        if ((vectorDimensions == null) == (modelId == null)) {
+            throw new IllegalArgumentException(
+                "[" + NAME + "] requires one of [" + MODEL_ID.getPreferredName() + "], or [" + VECTOR_DIMENSIONS.getPreferredName() + "]"
+            );
+        }
+        if (modelId != null && modelText == null) {
+            throw new IllegalArgumentException(
+                "[" + NAME + "] requires [" + MODEL_TEXT.getPreferredName() + "] when [" + MODEL_ID.getPreferredName() + "] is specified"
+            );
+        }
+
+        this.fieldName = fieldName;
+        this.modelText = modelText;
+        this.modelId = modelId;
+        this.vectorDimensions = vectorDimensions;
+        this.tokenPruningConfig = tokenPruningConfig;
+    }
+
+    public SparseVectorQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        this.fieldName = in.readString();
+        this.modelText = in.readString();
```

Comment:
If the `vector_dimensions` parameter is used won't both `modelId` and `modelText` be null?

Reply:
Indeed, modelId & modelText should be considered nullable. We are unifying the query builder here and removing the weighted tokens builder. So, we need to make sure users can optionally provide model text & id, or optionally provide vector dimensions.

Reply:
Good catch, thanks!